dev/RadioButtons/RadioButtons.cpp

```diff
@@ -7,136 +7,634 @@
+                
+            toggleButton.SetValue(s_childHandlersProperty, childHandlers.as<winrt::IInspectable>());
+        }
+        if (auto const repeater = m_repeater.get())
+        {
+            if (auto const itemSourceView = repeater.ItemsSourceView())
+            {
+                element.SetValue(winrt::AutomationProperties::PositionInSetProperty(), box_value(args.Index() + 1));
+                element.SetValue(winrt::AutomationProperties::SizeOfSetProperty(), box_value(itemSourceView.Count()));
+            }
+        }
     }
+}
 
-    m_selectionChangedEventSource(sender, args);
+void RadioButtons::OnRepeaterElementClearing(const winrt::ItemsRepeater&, const winrt::ItemsRepeaterElementClearingEventArgs& args)
+{
+    if (auto const element = args.Element())
+    {
+        if (auto const childHandlers = element.GetValue(s_childHandlersProperty).try_as<ChildHandlers>())
+        {
+            childHandlers->checkedRevoker.revoke();
+            childHandlers->uncheckedRevoker.revoke();
+        }
+    }
 }
 
-void RadioButtons::OnListViewKeyDown(const winrt::IInspectable& sender, const winrt::KeyRoutedEventArgs& args)
+void RadioButtons::OnRepeaterElementIndexChanged(const winrt::ItemsRepeater&, const winrt::ItemsRepeaterElementIndexChangedEventArgs& args)
 {
-    if (args.Key() == winrt::VirtualKey::Control)
+    if (auto const element = args.Element())
     {
-        m_isControlDown = true;
+        element.SetValue(winrt::AutomationProperties::PositionInSetProperty(), box_value(args.NewIndex() + 1));
     }
-    else if (args.Key() == winrt::VirtualKey::Down)
+}
+
+
+void RadioButtons::OnRepeaterCollectionChanged(const winrt::IInspectable&, const winrt::IInspectable&)
+{
+    if (auto const repeater = m_repeater.get())
     {
-        args.Handled(MoveSelection(1));
+        if (auto const itemSourceView = repeater.ItemsSourceView())
+        {
+            auto const count = itemSourceView.Count();
+            for (auto index = 0; index < count; index++)
+            {
+                if (auto const radioButton = repeater.TryGetElement(index))
+                {
+                    radioButton.SetValue(winrt::AutomationProperties::SizeOfSetProperty(), box_value(count));
+                }
+            }
+        }
     }
-    else if (args.Key() == winrt::VirtualKey::Up)
+}
+
+void RadioButtons::OnSelectionChanged(const winrt::IInspectable&, const winrt::SelectionModelSelectionChangedEventArgs& args)
+{
+    auto const currentIndex = m_selectionModel.SelectedIndex();
+    auto const repeater = m_repeater.get();
+
+    if (repeater)
     {
-        args.Handled(MoveSelection(-1));
+        if (auto const previousIndices = args.PreviousSelectedIndices())
+        {
+            for (auto const previousIndex : previousIndices)
+            {
+                if (previousIndex)
+                {
```

Comment:
These seem like a lot of checks. Should these be filtered out in PreviousSelectedIndices itself. If not, it might make it easier to group these checks into the same if statement.

Reply:
I think not, but I do like moving it into the same if statement.

---
In reply to: [347430041](https://github.com/microsoft/microsoft-ui-xaml/pull/1623#discussion_r347430041) [](ancestors = 347430041)